src/ray/raylet/scheduling/local_resource_manager.cc

```diff
@@ -500,12 +500,21 @@ ray::gcs::NodeResourceInfoAccessor::ResourceMap LocalResourceManager::GetResourc
   return map;
 }
 
-void LocalResourceManager::OnResourceChanged() {
+void LocalResourceManager::OnResourceOrStateChanged() {
+  if (IsLocalNodeDraining() && IsLocalNodeIdle()) {
+    // The node is drained.
+    // Sending a SIGTERM to itself is equivalent to gracefully shutting down raylet.
+    RAY_LOG(INFO) << "The node is drained, exiting...";
+    RAY_CHECK(std::raise(SIGTERM) == 0) << "There was a failure while sending a "
```

Comment:
nit: should we create a helper function to shutdown the raylet..?

Reply:
Sure. I can do it in a follow-up PR. We have several places in the codebase that do this.